### ISSUE TITLE
HDF5 dataset creation support

### ISSUE THREAD

**Issue by sandrain** [LLNL/UnifyCR#114]
### Testing environment
The test has been conducted in our testbed with three x86 nodes:
* Intel Xeon  E5-2603
* 64 GB RAM
* mpi/mpich-3.2-x86_64

The `write_test_static` seems to run fine.

### HDF5 dataset creation test
The basic hdf5 file creation is not working as it is supposed to be. The sequence of hdf5 calls are as follows:

```
H5Fcreate
H5Screate_simple
H5Dcreate2
H5Dclose
H5Sclose
H5Fclose
```

The full testing program was taken from [C Examples In HDF5 Source Code](https://support.hdfgroup.org/ftp/HDF5/examples/src-html/c.html), the create [example](https://support.hdfgroup.org/ftp/HDF5/current/src/unpacked/examples/h5_crtdat.c).
 It has been modified to use MPI as follows:

```
/*
 *  This example illustrates how to create a dataset that is a 4 x 6 
 *  array.  It is used in the HDF5 Tutorial.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <mpi.h>
#include "hdf5.h"
#define FILE "dset.h5"

int main(int argc, char **argv) {

    int rank, rank_num;

    hid_t       file_id, dataset_id, dataspace_id;  /* identifiers */
    hsize_t     dims[2];
    herr_t      status = 0;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &rank_num);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    unifycr_mount("/tmp", rank, rank_num, 0, 1);

    MPI_Barrier(MPI_COMM_WORLD);

    if (rank == 0) {
        /* Create a new file using default properties. */
        file_id = H5Fcreate(FILE, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
        printf("H5Fcreate: %d\n", file_id);

        /* Create the data space for the dataset. */
        dims[0] = 4; 
        dims[1] = 6; 
        dataspace_id = H5Screate_simple(2, dims, NULL);
        printf("H5Screate_simple: %d\n", dataspace_id);

        /* Create the dataset. */
        dataset_id = H5Dcreate2(file_id, "/tmp/dset", H5T_STD_I32BE,
                dataspace_id, H5P_DEFAULT, H5P_DEFAULT,
                H5P_DEFAULT);
        printf("H5Dcreate2: %d\n", status);

        /* End access to the dataset and release resources used by it. */
        status = H5Dclose(dataset_id);
        printf("H5Dclose: %d\n", status);

        /* Terminate access to the data space. */ 
        status = H5Sclose(dataspace_id);
        printf("H5Sclose: %d\n", status);

        /* Close the file. */
        status = H5Fclose(file_id);
        printf("H5Fclose: %d\n", status);
    }

    MPI_Barrier(MPI_COMM_WORLD);

    return 0;
}
```

The output of the program execution is:

```
unifycr: unifycr.c:1771: unifycr_init: are we using spillover? 1

unifycr: unifycr.c:1875: unifycr_init: FD limit for system = 1024

unifycr: unifycr.c:1444: unifycr_superblock_shmget: Key for superblock = 0

unifycr: unifycr.c:1771: unifycr_init: are we using spillover? 1

unifycr: unifycr.c:1771: unifycr_init: are we using spillover? 1

unifycr: unifycr.c:1875: unifycr_init: FD limit for system = 1024

unifycr: unifycr.c:1444: unifycr_superblock_shmget: Key for superblock = 0

unifycr: unifycr.c:1402: unifycr_init_structures: Meta-stacks initialized!

unifycr: unifycr.c:1402: unifycr_init_structures: Meta-stacks initialized!

unifycr: unifycr.c:1875: unifycr_init: FD limit for system = 1024

unifycr: unifycr.c:1444: unifycr_superblock_shmget: Key for superblock = 0

unifycr: unifycr.c:1402: unifycr_init_structures: Meta-stacks initialized!

unifycr: unifycr.c:668: unifycr_fid_alloc: unifycr_stack_pop() gave 0

unifycr: unifycr.c:704: unifycr_fid_create_file: Filename /tmp got unifycr fd 0

unifycr: unifycr.c:668: unifycr_fid_alloc: unifycr_stack_pop() gave 0

unifycr: unifycr.c:704: unifycr_fid_create_file: Filename /tmp got unifycr fd 0

unifycr: unifycr.c:668: unifycr_fid_alloc: unifycr_stack_pop() gave 0

unifycr: unifycr.c:704: unifycr_fid_create_file: Filename /tmp got unifycr fd 0

H5Fcreate: 16777216
H5Screate_simple: 67108866
HDF5-DIAG: Error detected in HDF5 (1.8.12) thread 0:
  #000: ../../src/H5D.c line 170 in H5Dcreate2(): unable to create dataset
    major: Dataset
    minor: Unable to initialize object
  #001: ../../src/H5Dint.c line 439 in H5D__create_named(): unable to create and link to dataset
    major: Dataset
    minor: Unable to initialize object
  #002: ../../src/H5L.c line 1638 in H5L_link_object(): unable to create new link to object
    major: Links
    minor: Unable to initialize object
  #003: ../../src/H5L.c line 1882 in H5L_create_real(): can't insert link
    major: Symbol table
    minor: Unable to insert object
  #004: ../../src/H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #005: ../../src/H5Gtraverse.c line 755 in H5G_traverse_real(): component not found
    major: Symbol table
    minor: Object not found
H5Dcreate2: 0
HDF5-DIAG: Error detected in HDF5 (1.8.12) thread 0:
  #000: ../../src/H5D.c line 391 in H5Dclose(): not a dataset
    major: Invalid arguments to routine
    minor: Inappropriate type
H5Dclose: -1
H5Sclose: 0
H5Fclose: 0
```
It seems like `H5Dcreate2()` function fails when it tries to create an named object. Further debugging would necessary to figure out what exactly is problematic (`lstat`, `lseek` or `write`, based on the `strace` result below).

UnifyCR was configured as follows:

```
$ env | grep UNIFYCR
UNIFYCR_DEBUG=8
UNIFYCR_META_DB_NAME=unifycr_db
UNIFYCR_SERVER_DEBUG_LOG=/tmp/unifycr/unifycrd_debug.31028
UNIFYCR_EXTERNAL_META_DIR=/tmp/unifycr/ssd
UNIFYCR_META_SERVER_RATIO=1
UNIFYCR_EXTERNAL_DATA_DIR=/tmp/unifycr/ssd
UNIFYCR_CHUNK_MEM=0
UNIFYCR_META_DB_PATH=/tmp/unifycr/ssd
```

The original example (without modification) works like:

```bash
$ cc -o h5_create_dataset h5_create_dataset.c -lhdf5
$ ./h5_create_dataset
$ h5dump dset.h5
HDF5 "dset.h5" {
GROUP "/" {
   DATASET "dset" {
      DATATYPE  H5T_STD_I32BE
      DATASPACE  SIMPLE { ( 4, 6 ) / ( 4, 6 ) }
      DATA {
      (0,0): 0, 0, 0, 0, 0, 0,
      (1,0): 0, 0, 0, 0, 0, 0,
      (2,0): 0, 0, 0, 0, 0, 0,
      (3,0): 0, 0, 0, 0, 0, 0
      }
   }
}
}
$ strace ./h5_create_dataset
...
open("dset.h5", O_RDWR)                 = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=1400, ...}) = 0
close(3)                                = 0
open("dset.h5", O_RDWR|O_CREAT|O_TRUNC, 0666) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=0, ...}) = 0
mmap(NULL, 528384, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fb7c0536000
getcwd("/autofs/nccs-svm1_techint/home/hs2/projects/__tmp/h5", 1024) = 53
lstat("dset.h5", {st_mode=S_IFREG|0644, st_size=0, ...}) = 0
lseek(3, 0, SEEK_SET)                   = 0
write(3, "\211HDF\r\n\32\n\0\0\0\0\0\10\10\0\4\0\20\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 96) = 96
write(3, "\1\0\1\0\1\0\0\0\30\0\0\0\0\0\0\0\21\0\20\0\0\0\0\0\210\0\0\0\0\0\0\0"..., 1304) = 1304
munmap(0x7fb7c0536000, 528384)          = 0
close(3)                                = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

Basically, it uses the following system calls:

```
open
fstat
close
lstat
lseek
write
close
```

**Comment by fwang2** [comment]
@sandrain 
this is good, exactly what we need (to kickstart the integration). Among the system calls you have traced, which ones have existing support or if not, are low hanging fruits? and which set are more involved or difficult to get it done?

**Comment by sandrain** [comment]
I guess the first thing we need to do is list the system and library calls that UnifyCR is supposed to support and then fully implement each of them one by one.

**Comment by sandrain** [comment]
The following is the some more debugging info (with printf):

```
...
write(1, "### calling: H5Fcreate\n", 23### calling: H5Fcreate
) = 23
brk(NULL)                               = 0x1352000
brk(0x1373000)                          = 0x1373000
brk(NULL)                               = 0x1373000
brk(NULL)                               = 0x1373000
brk(0x1394000)                          = 0x1394000
open("dset.h5", O_RDWR)                 = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=1400, ...}) = 0
close(3)                                = 0
open("dset.h5", O_RDWR|O_CREAT|O_TRUNC, 0666) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=0, ...}) = 0
mmap(NULL, 528384, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ff703498000
getcwd("/autofs/nccs-svm1_techint/home/hs2/projects/UnifyCR/client/tests", 1024) = 65
lstat("dset.h5", {st_mode=S_IFREG|0644, st_size=0, ...}) = 0
write(1, "\n", 1
)                       = 1
write(1, "### calling: H5Screate_simple\n", 30### calling: H5Screate_simple
) = 30
write(1, "\n", 1
)                       = 1
write(1, "### calling: H5Dcreate2\n", 24### calling: H5Dcreate2
) = 24
write(1, "\n", 1
)                       = 1
write(1, "### calling: H5Dclose\n", 22### calling: H5Dclose
) = 22
write(1, "\n", 1
)                       = 1
write(1, "### calling: H5Sclose\n", 22### calling: H5Sclose
) = 22
write(1, "\n", 1
)                       = 1
write(1, "### calling: H5Fclose\n", 22### calling: H5Fclose
) = 22
lseek(3, 0, SEEK_SET)                   = 0
write(3, "\211HDF\r\n\32\n\0\0\0\0\0\10\10\0\4\0\20\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 96) = 96
write(3, "\1\0\1\0\1\0\0\0\30\0\0\0\0\0\0\0\21\0\20\0\0\0\0\0\210\0\0\0\0\0\0\0"..., 1304) = 1304
munmap(0x7ff703498000, 528384)          = 0
close(3)                                = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

It looks like I need to further debugging/investigation, because the test program does not invoke I/O system calls after creating the file (H5Fcreate, calling open(2), fstat(2) and lstat(2)), until the file is closed (H5Fclose, calling write(2)). The write(2) calls are from printf statements that I inserted for debugging.

The following is the full source of the program (not a MPI program):

```
#include "hdf5.h"
#define FILE "dset.h5"

int main() {

   hid_t       file_id, dataset_id, dataspace_id;  /* identifiers */
   hsize_t     dims[2];
   herr_t      status;

   printf("### calling: H5Fcreate\n");

   /* Create a new file using default properties. */
   file_id = H5Fcreate(FILE, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);

   printf("\n### calling: H5Screate_simple\n");

   /* Create the data space for the dataset. */
   dims[0] = 4; 
   dims[1] = 6; 
   dataspace_id = H5Screate_simple(2, dims, NULL);

   printf("\n### calling: H5Dcreate2\n");

   /* Create the dataset. */
   dataset_id = H5Dcreate2(file_id, "/dset", H5T_STD_I32BE, dataspace_id, 
                          H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
   printf("\n### calling: H5Dclose\n");

   /* End access to the dataset and release resources used by it. */
   status = H5Dclose(dataset_id);

   printf("\n### calling: H5Sclose\n");

   /* Terminate access to the data space. */ 
   status = H5Sclose(dataspace_id);

   printf("\n### calling: H5Fclose\n");

   /* Close the file. */
   status = H5Fclose(file_id);
}
```

**Comment by sandrain** [comment]
I went through the debugging, and it turned out that the first failing point was a missing system call of fxstat. 

```
(gdb) bt
#0  H5E_printf_stack (estack=0x0, file=0x7f79bbfd4fe1 "H5FDsec2.c", func=0x7f79bbfd56b0 <__func__.10535> "H5FD_sec2_open", line=347, 
    cls_id=184549376, maj_id=201326595, min_id=201326683, fmt=0x7f79bbfd5210 "%s, errno = %d, error message = '%s'") at H5Eint.c:705
#1  0x00007f79bbc63aab in H5FD_sec2_open (name=0x400f05 "/tmp/dset.h5", flags=19, fapl_id=167772165, maxaddr=9223372036854775807) at H5FDsec2.c:347
#2  0x00007f79bbc4d54c in H5FD_open (name=0x400f05 "/tmp/dset.h5", flags=19, fapl_id=167772165, maxaddr=9223372036854775807) at H5FD.c:986
#3  0x00007f79bbc2c8ce in H5F_open (name=0x400f05 "/tmp/dset.h5", flags=19, fcpl_id=167772164, fapl_id=167772165, dxpl_id=167772168) at H5F.c:1281
#4  0x00007f79bbc2da3e in H5Fcreate (filename=0x400f05 "/tmp/dset.h5", flags=19, fcpl_id=167772164, fapl_id=167772165) at H5F.c:1502
#5  0x0000000000400d29 in main (argc=1, argv=0x7ffff77e8008) at test_hdf5_create_dataset.c:51
(gdb) frame 1
#1  0x00007f79bbc63aab in H5FD_sec2_open (name=0x400f05 "/tmp/dset.h5", flags=19, fapl_id=167772165, maxaddr=9223372036854775807) at H5FDsec2.c:347
347             HSYS_GOTO_ERROR(H5E_FILE, H5E_BADFILE, NULL, "unable to fstat file")
```

Basically, `H5Fcreate` internally calls the underlying [virtual file driver](https://support.hdfgroup.org/HDF5/doc/Supplements/stackvfd/Stackable_VFD.pdf)'s implementation of `open` and `stat`. The default I/O driver (`sec2`).

The following is the output from the unifycr code:

```

unifycr: unifycr.c:1725: unifycr_init: gotcha_wrap returned 1

unifycr: unifycr.c:1732: unifycr_init: This function name failed to be wrapped: stat

unifycr: unifycr.c:1771: unifycr_init: are we using spillover? 1

unifycr: unifycr.c:1875: unifycr_init: FD limit for system = 1024

unifycr: unifycr.c:1444: unifycr_superblock_shmget: Key for superblock = 0

unifycr: unifycr.c:1725: unifycr_init: gotcha_wrap returned 1

unifycr: unifycr.c:1732: unifycr_init: This function name failed to be wrapped: stat

unifycr: unifycr.c:1771: unifycr_init: are we using spillover? 1

unifycr: unifycr.c:1875: unifycr_init: FD limit for system = 1024

unifycr: unifycr.c:1402: unifycr_init_structures: Meta-stacks initialized!

unifycr: unifycr.c:1444: unifycr_superblock_shmget: Key for superblock = 0

unifycr: unifycr.c:1402: unifycr_init_structures: Meta-stacks initialized!

unifycr: unifycr.c:1725: unifycr_init: gotcha_wrap returned 1

unifycr: unifycr.c:1732: unifycr_init: This function name failed to be wrapped: stat

unifycr: unifycr.c:1771: unifycr_init: are we using spillover? 1

unifycr: unifycr.c:1875: unifycr_init: FD limit for system = 1024

unifycr: unifycr.c:1444: unifycr_superblock_shmget: Key for superblock = 0

unifycr: unifycr.c:668: unifycr_fid_alloc: unifycr_stack_pop() gave 0

unifycr: unifycr.c:704: unifycr_fid_create_file: Filename /tmp got unifycr fd 0

unifycr: unifycr.c:668: unifycr_fid_alloc: unifycr_stack_pop() gave 0

unifycr: unifycr.c:704: unifycr_fid_create_file: Filename /tmp got unifycr fd 0

unifycr: unifycr.c:1402: unifycr_init_structures: Meta-stacks initialized!

unifycr: unifycr.c:668: unifycr_fid_alloc: unifycr_stack_pop() gave 0

unifycr: unifycr.c:704: unifycr_fid_create_file: Filename /tmp got unifycr fd 0

unifycr: unifycr.c:1135: unifycr_fid_open: unifycr_get_fid_from_path() gave -1

unifycr: unifycr.c:1224: unifycr_fid_open: Couldn't find entry for /tmp/dset.h5 in UNIFYCR

unifycr: unifycr.c:1135: unifycr_fid_open: unifycr_get_fid_from_path() gave -1

unifycr: unifycr.c:1188: unifycr_fid_open: Couldn't find entry for /tmp/dset.h5 in UNIFYCR

unifycr: unifycr.c:1192: unifycr_fid_open: unifycr_superblock = 0x7f79b19b2000; free_fid_stack = 0x7f79b19b2004;free_chunk_stack = 0x7f79b19d7c0c; unifycr_filelist = 0x7f79b19b220c;chunks = 0x7f79b19d8000

unifycr: unifycr.c:668: unifycr_fid_alloc: unifycr_stack_pop() gave 1

unifycr: unifycr.c:704: unifycr_fid_create_file: Filename /tmp/dset.h5 got unifycr fd 1

unifycr: unifycr.c:1262: unifycr_fid_open: UNIFYCR_open generated fd 1 for file /tmp/dset.h5

unifycr: unifycr-sysio.c:645: __wrap_open: UNIFYCR_open generated fd 1 for file /tmp/dset.h5

unifycr: unifycr.c:462: unifycr_intercept_fd: Changing fd from exposed 1025 to internal 1

Function not yet supported @ unifycr-sysio.c:1811

```

The current `fxstat` implementation also acts confusing, since it returns `EBADF` instead of `ENOSYS`. I will create a separate issue for the implementation of `fxstat` and probably other `stat` family calls. After that, I will come back here and continue the testing for any further limitations.

**Comment by sandrain** [comment]
After tentatively implementing `fxstat(2)`, it now fails from missing `lxstat(2)`. These are all from `H5Fcreate` function which create a new HDF file. I will implement all missing functions then submit a PR.

**Comment by sandrain** [comment]
Now, the dataset creation program is working. In addition to the missing syscalls (`fxstat(2)`, `lxstat(2)`, and `lseek64(2)`), there was a bug in the `__wrap_write()` implementation where if condition was flipped. See [here](https://github.com/sandrain/UnifyCR/commit/6dff68c44c39a170729c6b22d3acb67087a64bf4).

I have moved on to the reading/writing datasets. It fails when looking for the existing dataset (the dataset which has been created by the create example program). Specifically, it can successfully open the file with `H5Fopen()`, but fails during the subsequent `H5Dopen2()`.

I am not sure if I need to submit the PR with the current progress or submit it later after finishing/fixing the read/write tests as well.

**Comment by sandrain** [comment]
As @kathrynmohror suggested, I will separate patches and submit PRs one by one. It's going to be:

* Implementation of `fstat(2)` related syscalls
* Implementation of `lseek(2)` related syscalls
* Fix of `write(2)` syscall
* HDF5 testing programs and build script for them

**Comment by sandrain** [comment]
This has been addressed by #126 and #128.
